packing.go

```diff
@@ -30,50 +31,96 @@ func Pack(source interface{}) (map[string]interface{}, error) {
+
+//MsgEnvelopesToMessages converts envelopes to any number of different structs that impliment the Message interface
+func MsgEnvelopesToMessages(msgEnvelopes []*repository.MessageEnvelope, converters ...MessageConverter) []Message {
+	myConverters := append(converters, defaultConverters()...)
+
 	messages := make([]Message, 0, len(msgEnvelopes))
 	for _, messageEnvelope := range msgEnvelopes {
 		if messageEnvelope == nil {
 			logrus.Error("Found a nil in the message envelope slice, can't transform to a message")
 			continue
 		}
-		data := make(map[string]interface{})
-		err := json.Unmarshal(messageEnvelope.Data, &data)
-		if err != nil {
-			logrus.WithError(err).Error("Can't unmarshal JSON from message envelope")
-			continue
-		}
-		if strings.HasSuffix(messageEnvelope.Stream, ":command") {
-			command := &Command{
-				NewID:      messageEnvelope.MessageID,
-				Type:       messageEnvelope.Type,
-				Category:   strings.TrimSuffix(messageEnvelope.Stream, ":command"),
-				CausedByID: messageEnvelope.CausedByID,
-				OwnerID:    messageEnvelope.OwnerID,
-				Data:       data,
-			}
-			messages = append(messages, command)
-		} else {
-			category, id := "", ""
-			cats := strings.SplitN(messageEnvelope.Stream, "-", 2)
-			if len(cats) > 0 {
-				category = cats[0]
-				if len(cats) == 2 {
-					id = cats[1]
-				}
-			}
-			event := &Event{
-				NewID:      messageEnvelope.MessageID,
-				Type:       messageEnvelope.Type,
-				Category:   category,
-				CategoryID: id,
-				CausedByID: messageEnvelope.CausedByID,
-				OwnerID:    messageEnvelope.OwnerID,
-				Data:       data,
+
+		for _, converter := range myConverters {
+			message, err := converter(messageEnvelope)
+			if message != nil && err == nil {
+				messages = append(messages, message)
+				break // only one successful conversion per envelope
 			}
-			messages = append(messages, event)
+
 		}
 	}
 
 	return messages
 }
+
+func convertEnvelopeToCommand(messageEnvelope *repository.MessageEnvelope) (Message, error) {
+	if strings.HasSuffix(messageEnvelope.StreamName, ":command") {
+		data := make(map[string]interface{})
+		if err := json.Unmarshal(messageEnvelope.Data, &data); err != nil {
+			logrus.WithError(err).Error("Can't unmarshal JSON from message envelope data")
+		}
+		metadata := make(map[string]interface{})
+		if err := json.Unmarshal(messageEnvelope.Metadata, &metadata); err != nil {
+			logrus.WithError(err).Error("Can't unmarshal JSON from message envelope metadata")
+		}
+		command := &Command{
+			ID:             messageEnvelope.ID,
+			MessageType:    messageEnvelope.MessageType,
+			StreamCategory: strings.TrimSuffix(messageEnvelope.StreamName, ":command"),
+			Version:        messageEnvelope.Version,
+			GlobalPosition: messageEnvelope.GlobalPosition,
+			Data:           data,
+			Metadata:       metadata,
+			Time:           messageEnvelope.Time,
+		}
+
+		return command, nil
+	} else {
+		return nil, errors.New("not a command, also, this isn't an error")
```

Comment:
Attempted to run convertEnvelopeToCommand for a StreamName that does not contain a ":command" suffix.